tchannel/testing/vcr/cassette.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import
+
+from itertools import chain
+from collections import deque
+
+from . import types
+from . import yaml
+from .exceptions import RequestNotFoundError, UnsupportedVersionError
+
+
+# Version of the storage format.
+VERSION = 1
+
+
+class Cassette(object):
+    """Represents a series of recorded interactions."""
+
+    def __init__(self, path):
+        """Initialize a new cassette.
+
+        :param path:
+            File path at which this cassette will be stored.
+        """
+        self.path = path
+
+        # Number of times interactions from this cassette have been played.
+        self.play_count = 0
+
+        self._available = deque()
+        self._played = deque()
```

Comment:
Whether an interaction is played or not is knowledge limited/known only for one use of the cassette. When you load a cassette, they're all available in that they can all be replayed. As you play some of them, they're moved to the `played` deque because they're no longer available to be played -- you don't want to allow duplicate plays of the same interaction. During that session, if you record more interactions, they're added to the `recorded` deque because they're also not available to be replayed -- you don't want to prevent someone from making a request twice if they need to, especially if they're testing some retry logic. When you save the cassette, you want all these interactions to be saved--those that were played, those that were not played, and those that were recorded. After that, everything is available because that session with the cassette is done.